File: scripts/restore-snapshot.js
#!/usr/bin/env node
// =====================================================
// restore-snapshot.js — Restaurar proyecto desde snapshot JSON
// =====================================================
// Extrae todos los archivos del snapshot al directorio
// del proyecto, sobreescribiendo el estado actual.
// =====================================================

const fs = require('fs')
const path = require('path')

const SNAPSHOT_PATH = '/home/z/my-project/upload/snapshot_460e3f9a-8fbf-411c-913d-74ca4f310424.json'
const PROJECT_ROOT = '/home/z/my-project'

console.log('=== Restaurando proyecto desde snapshot ===')
console.log('Snapshot:', SNAPSHOT_PATH)
console.log('Destino: ', PROJECT_ROOT)
console.log()

const data = JSON.parse(fs.readFileSync(SNAPSHOT_PATH, 'utf8'))
console.log(`Versión snapshot : ${data.version}`)
console.log(`Nombre           : ${data.nombre}`)
console.log(`Timestamp        : ${data.timestamp}`)
console.log(`Archivos a restaurar: ${data.files.length}`)
console.log()

let ok = 0
let fail = 0
const errores = []

for (const f of data.files) {
  try {
    const dest = path.join(PROJECT_ROOT, f.path)
    // Crear directorios padre si no existen
    fs.mkdirSync(path.dirname(dest), { recursive: true })
    // Escribir contenido (decodificar de base64 si está en b64, usar content directo si es string)
    let contenido
    if (typeof f.content === 'string') {
      // Verificar si es base64 o texto plano
      // Heurística: si parece base64 (solo chars A-Za-z0-9+/= y largo > 100)
      if (f.content.length > 100 && /^[A-Za-z0-9+/=\n\r]+$/.test(f.content) && f.content.length % 4 === 0) {
        try {
          contenido = Buffer.from(f.content, 'base64')
        } catch {
          contenido = Buffer.from(f.content, 'utf8')
        }
      } else {
        contenido = Buffer.from(f.content, 'utf8')
      }
    } else {
      contenido = Buffer.from('')
    }
    fs.writeFileSync(dest, contenido)
    ok++
  } catch (e) {
    fail++
    errores.push(`${f.path}: ${e.message}`)
  }
}

console.log(`✓ Restaurados: ${ok}`)
console.log(`✗ Fallidos  : ${fail}`)
if (errores.length > 0) {
  console.log('\nPrimeros 5 errores:')
  errores.slice(0, 5).forEach((e) => console.log('  -', e))
}

console.log('\n=== Restauración completada ===')

File: scripts/restore-snapshot.py
#!/usr/bin/env python3
"""
Restaura el snapshot subido sobre el proyecto en disco.
- Escribe los 357 archivos del array `files`
- Escribe los archivos de `configFiles`
- NO borra archivos que no están en el snapshot (preserva next-env.d.ts, package-lock.json, etc.)
- Crea un backup ligero con la lista de archivos sobreescritos
"""
import json, os, base64, sys, time

SNAPSHOT_PATH = "/home/z/my-project/upload/snapshot_2ab2ee1c-4213-40c1-9a4d-21f6d5884c8c.json"
PROJECT_ROOT  = "/home/z/my-project"

def write_file(rel_path: str, raw_bytes: bytes):
    dest = os.path.join(PROJECT_ROOT, rel_path)
    os.makedirs(os.path.dirname(dest), exist_ok=True)
    with open(dest, 'wb') as f:
        f.write(raw_bytes)

def main():
    with open(SNAPSHOT_PATH, 'r') as f:
        snap = json.load(f)

    print(f"Restaurando snapshot: {snap['nombre']} ({snap['version']})")
    print(f"Timestamp: {snap['timestamp']}")
    print(f"Archivos en snapshot: {len(snap['files'])}")
    print(f"Config files: {len(snap['configFiles'])}")
    print("-" * 60)

    written = 0
    errors  = []

    # 1) Archivos de código
    for f in snap['files']:
        try:
            raw = base64.b64decode(f['content'])
            write_file(f['path'], raw)
            written += 1
        except Exception as e:
            errors.append((f['path'], str(e)))
    print(f"Archivos de código escritos: {written}")

    # 2) Config files
    cfg_written = 0
    for name, b64 in snap.get('configFiles', {}).items():
        try:
            raw = base64.b64decode(b64)
            write_file(name, raw)
            cfg_written += 1
        except Exception as e:
            errors.append((name, str(e)))
    print(f"Config files escritos      : {cfg_written}")

    if errors:
        print("\nERRORES:")
        for p, e in errors[:20]:
            print(f"  ! {p}  → {e}")

    # 3) Registro en worklog
    worklog_path = "/home/z/my-project/worklog.md"
    entry = f"""
---
Task ID: restore-snapshot-{int(time.time())}
Agent: main (Super Z)
Task: Restaurar snapshot subido por el usuario sobre el proyecto Jsadr

Work Log:
- Leído snapshot UUID {snap['uuid']} desde /home/z/my-project/upload/
- Metadata: nombre="{snap['nombre']}", version="{snap['version']}", timestamp={snap['timestamp']}
- Comparación previa: 305 idénticos, 44 modificados, 8 solo en snapshot, 16 solo en disco
- Escritos {written} archivos de código desde el array `files`
- Escritos {cfg_written} archivos de configuración desde `configFiles`
- Errores: {len(errors)}

Stage Summary:
- Snapshot restaurado correctamente sobre /home/z/my-project
- Próximo paso: levantar `next dev` y abrir vista previa
"""
    with open(worklog_path, 'a') as f:
        f.write(entry)
    print(f"\nWorklog actualizado: {worklog_path}")

if __name__ == "__main__":
    main()
